### DISCUSSION TITLE
JS Framework Benchmark Implementation

### DISCUSSION THREAD

**Discussion by ryansolid** [alpinejs/alpine#822] [4. General]
I've been working on a JS Framework Benchmark implementation: https://github.com/krausest/js-framework-benchmark/pull/806

The library seems pretty easy to use and only took me a matter of minutes to whip this together and it just seemed to work. I just would appreciate some eyes to make sure I'm not doing anything really unoptimal. Maybe some small performance improvements I might not be aware of.

Given the nature of the test I've used npm and a bundler to minify the code. Otherwise tried to keep things as idiomatic as possible. I noticed that it isn't treeshakable and there is another issue for that but I suppose given most use is from a CDN the non-treeshaked approach is probably still the most representative. But let me know, as I want to best represent the library.

I have witnessed some strange behavior around it being run from the chromedriver where clicks aren't being registered when the tests are actually running. I haven't hit this before in other implementations I've done for this benchmark so if you had any insights that would be amazing too.

I will continue to try to debug this. But I'd appreciate any insight you can give. Thank you.

**Comment by ryansolid** [comment]
Haven't really found anything yet. Beyond the chrome driver issues, I'd love to sanity check I'm not doing anything terribly wrong.  I'm getting slower numbers than I'd hope. But this is all it is. A really simple TODOs style app. The JS is getting loaded from NPM and being minified and bundled with rollup+terser. Any help very much appreciated.

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Alpine-keyed</title>
    <link href="/css/currentStyle.css" rel="stylesheet" />
  </head>
  <body>
    <div class="container" id="main" x-data="app()">
      <div class="jumbotron">
        <div class="row">
          <div class="col-md-6">
            <h1>Alpine-keyed</h1>
          </div>
          <div class="col-md-6">
            <div class="row">
              <div class="col-sm-6 smallpad">
                <button
                  type="button"
                  class="btn btn-primary btn-block"
                  id="run"
                  @click="run"
                >
                  Create 1,000 rows
                </button>
              </div>
              <div class="col-sm-6 smallpad">
                <button
                  type="button"
                  class="btn btn-primary btn-block"
                  id="runlots"
                  @click="runLots"
                >
                  Create 10,000 rows
                </button>
              </div>
              <div class="col-sm-6 smallpad">
                <button
                  type="button"
                  class="btn btn-primary btn-block"
                  id="add"
                  @click="add"
                >
                  Append 1,000 rows
                </button>
              </div>
              <div class="col-sm-6 smallpad">
                <button
                  type="button"
                  class="btn btn-primary btn-block"
                  id="update"
                  @click="update"
                >
                  Update every 10th row
                </button>
              </div>
              <div class="col-sm-6 smallpad">
                <button
                  type="button"
                  class="btn btn-primary btn-block"
                  id="clear"
                  @click="clear"
                >
                  Clear
                </button>
              </div>
              <div class="col-sm-6 smallpad">
                <button
                  type="button"
                  class="btn btn-primary btn-block"
                  id="swaprows"
                  @click="swapRows"
                >
                  Swap Rows
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <table class="table table-hover table-striped test-data">
        <tbody>
          <template x-for="item of data" :key="item.id">
            <tr :class="item.id === selected ? 'danger' : ''">
              <td class="col-md-1" x-text="item.id"></td>
              <td class="col-md-4">
                <a @click="select(item.id)" x-text="item.label"></a>
              </td>
              <td class="col-md-1">
                <a @click="remove(item.id)"
                  ><span
                    class="glyphicon glyphicon-remove"
                    aria-hidden="true"
                  ></span
                ></a>
              </td>
              <td class="col-md-6"></td>
            </tr>
          </template>
        </tbody>
      </table>

      <span
        class="preloadicon glyphicon glyphicon-remove"
        aria-hidden="true"
      ></span>
    </div>

    <script src="dist/main.js"></script>
  </body>
</html>
```

main.js
```js
import "alpinejs";

let idCounter = 1;
const adjectives = ["pretty", "large", "big", "small", "tall", "short", "long", "handsome", "plain", "quaint", "clean", "elegant", "easy", "angry", "crazy", "helpful", "mushy", "odd", "unsightly", "adorable", "important", "inexpensive", "cheap", "expensive", "fancy"],
  colours = ["red", "yellow", "blue", "green", "pink", "brown", "purple", "brown", "white", "black", "orange"],
  nouns = ["table", "chair", "house", "bbq", "desk", "car", "pony", "cookie", "sandwich", "burger", "pizza", "mouse", "keyboard"];

function _random (max) { return Math.round(Math.random() * 1000) % max; };

function buildData(count) {
  let data = new Array(count);
  for (let i = 0; i < count; i++) {
    data[i] = {
      id: idCounter++,
      label: `${adjectives[_random(adjectives.length)]} ${colours[_random(colours.length)]} ${nouns[_random(nouns.length)]}`
    }
  }
  return data;
}

window.app = function() {
  return {
    data: [],
    selected: undefined,

    add() { this.data = this.data.concat(buildData(1000)) },
    clear() {
      this.data = [];
      this.selected = undefined;
    },
    update() {
      for (let i = 0; i < this.data.length; i += 10) {
        this.data[i].label += ' !!!';
      }
    },
    remove(id) {
      const idx = this.data.findIndex(d => d.id === id);
      this.data.splice(idx, 1);
    },
    run() {
      this.data = buildData(1000);
      this.selected = undefined;
    },
    runLots() {
      this.data = buildData(10000);
      this.selected = undefined;
    },
    select(id) { this.selected = id },
    swapRows() {
      const d = this.data;
      if (d.length > 998) {
        const tmp = d[998];
        d[998] = d[1];
        d[1] = tmp;
      }
    }
  }
} 
```

**Reply by SimoTod** [reply]
I don't think you are doing anything wrong. Alpine's philosophy is about having a lot of small components not a single component containing the whole page. If you do that, yeah, it's not really performant. It's not meant to handle a x-for with thousands of nodes either, it will never "shine" with those banchmarks.

**Reply by ryansolid** [reply]
There is no `componentShouldUpdate` type of thing here I can leverage right? No benefit of nesting x-data bindings etc? The reactive system already knows to track what has changed so it already optimizes the update case. I always want to be as thorough as I can when writing these.

I understand this isn't the use case. I used KnockoutJS this way for years on initially server-rendered things. I understand the intention isn't to scale up this way and more scale-out. I will say this benchmark is more fair than ones that would render large data snapshots with the expectation of diffing them as this one uses only mouseevents to track like common things a user would do like selecting or moving items in a list one at a time. But it is a bit of hyperbole in terms of data size.

I've been just curious how these just use a CDN tag sprinkle JS libraries actually compare to tool optimized systems on size and raw performance. I think I will be doing a couple others as well like StimulusJS if possible. Having gone to the darkside mostly the last decade, when I see something like Alpine it's both incredibly nostalgic and sort of refreshing at the same time. But it also encourages me to me verify everything I've learned in the last 8 years.

**Reply by SimoTod** [reply]
No `componentShouldUpdate` I'm aware of. The reactivity layer doesn't track the changes, it refreshes the whole component (anything inside a tag with x-data but only that) when something changes in that component. This is the main reason why big components perform worse then small one. V3 should fix this scalability issue but there's no release date yet.
If you need more technical details, you should probably ask the author directly so you will get the right info from the source.

**Reply by ryansolid** [reply]
Thank you, That gives me one other angle to try, But I feel you have answered my questions quite well.

**Comment by ryansolid** [comment]
For anyone interested the Chromedriver issue wasn't Alpine's issue and the implementation was successfully merged: https://github.com/krausest/js-framework-benchmark/pull/806
